activerecord/lib/active_record/tasks/database_tasks.rb

```diff
@@ -268,13 +268,13 @@ def truncate_all(environment = env)
         end
       end
 
-      def migrate
+      def migrate(version = nil)
         check_target_version
 
         scope = ENV["SCOPE"]
         verbose_was, Migration.verbose = Migration.verbose, verbose?
 
-        Base.connection.migration_context.migrate(target_version) do |migration|
+        Base.connection.migration_context.migrate(target_version || version) do |migration|
```

Comment:
tests are failing and I believe this is because this line.
Maybe it's possible to patch interval (to grab all the versions before the `ENV['version']`

@eileencodes maybe share your idea so after that it will be easier to solve the issues and provide the tests if needed :)

Reply:
Hey sorry for the delay! This is what I was thinking in terms of an approach for fixing this issue. https://github.com/rails/rails/compare/main...eileencodes:fix-migration-ordering-bug?expand=1 It's similar to yours but avoids the complexity of connection_intervals. Let me know what you think. If you don't have time to pick this back up and write tests I can take this and make sure you get credit on the final PR.

Reply:
Yeah, that looks good, we can go both ways I can try to patch tests on my end but if you feel like that will be easier to do on your side I am not against it :)

My main goal is to fix migration order in the end

Reply:
I updated this request with your commit, will see how it goes with the tests

Reply:
The only question I have is `ActiveRecord::Base.establish_connection(db_config)` we call this for each migration

if we have a situation like (A and B here are databases):

mig1 from A, mig2 from A, mig3 from A, mig4 from A, mig1 from B, mig5 from A
do we really need to re-establish connections for the first 4 ?

I understand that this is just migrations and the optimization won't be that big but the easy check can be done if it's needed e.g.:


```
prev_config = db_config_by_version[1]
db_config_by_version.sort.each do |version, db_config|
      ActiveRecord::Base.establish_connection(db_config) unless prev_config == db_config
      ActiveRecord::Tasks::DatabaseTasks.migrate(version)
      prev_config = db_config
end
```

What do you think @eileencodes ?